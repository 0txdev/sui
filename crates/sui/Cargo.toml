[package]
name = "sui"
version.workspace = true
authors = ["Mysten Labs <build@mystenlabs.com>"]
license = "Apache-2.0"
publish = false
edition = "2021"

[dependencies]
json_to_table.workspace = true
tabled.workspace = true
anemo.workspace = true
anyhow.workspace = true
serde.workspace = true
serde_json.workspace = true
serde_yaml.workspace = true
signature.workspace = true
camino.workspace = true
tokio = { workspace = true, features = ["full"] }
async-trait.workspace = true
tracing.workspace = true
bcs.workspace = true
clap.workspace = true
bip32.workspace = true
rand.workspace = true
tap.workspace = true
inquire.workspace = true
rusoto_core.workspace = true
rusoto_kms.workspace = true
prometheus.workspace = true
git-version.workspace = true
const-str.workspace = true
num-bigint.workspace = true
regex.workspace = true
reqwest.workspace = true
im.workspace = true

sui-config.workspace = true
sui-execution = { path = "../../sui-execution" }
sui-swarm-config.workspace = true
sui-genesis-builder.workspace = true
sui-types.workspace = true
sui-json.workspace = true
sui-swarm.workspace = true
sui-json-rpc-types.workspace = true
sui-sdk.workspace = true
sui-keys.workspace = true
sui-source-validation.workspace = true
sui-move = { workspace = true, features = ["all"] }
sui-move-build.workspace = true
sui-protocol-config.workspace = true
shared-crypto.workspace = true
sui-replay.workspace = true

fastcrypto.workspace = true
fastcrypto-zkp.workspace = true

rustyline.workspace = true
rustyline-derive.workspace = true
colored.workspace = true
unescape.workspace = true
shell-words.workspace = true

tempfile.workspace = true
telemetry-subscribers.workspace = true

move-core-types.workspace = true
move-package.workspace = true
csv.workspace = true
workspace-hack.workspace = true

[target.'cfg(not(target_env = "msvc"))'.dependencies]
jemalloc-ctl.workspace = true

move-vm-test-utils = {workspace = true, optional = true}
move-vm-types = {workspace = true, optional = true}
move-vm-profiler = {workspace = true, optional = true}
move-cli = {workspace = true, optional = true}
move-unit-test = {workspace = true, optional = true}
move-vm-config = {workspace = true, optional = true}
sui-execution = {workspace = true, optional = true}
#move-vm-runtime = { path = "../../external-crates/move/move-execution/next-vm/crates/move-vm-runtime", option = true}

[dev-dependencies]
prometheus.workspace = true
fs_extra.workspace = true
expect-test.workspace = true
assert_cmd.workspace = true

test-cluster.workspace = true
sui-macros.workspace = true
sui-simulator.workspace = true
sui-test-transaction-builder.workspace = true

[package.metadata.cargo-udeps.ignore]
normal = ["jemalloc-ctl"]

[[example]]
name = "generate-genesis-checkpoint"
path = "src/generate_genesis_checkpoint.rs"
test = false

[[bin]]
name = "sui"
path = "src/main.rs"

[features]
default = ["gas-profiler"]
gas-profiler = [
    "sui-types/gas-profiler",
    "move-cli/gas-profiler",
    "move-unit-test/gas-profiler",
    "move-vm-config/gas-profiler",
    "move-vm-profiler/gas-profiler",
    "move-vm-types/gas-profiler",
    "move-vm-test-utils/gas-profiler",
    "sui-execution/gas-profiler",
    #"move-vm-runtime/gas-profiler",
]
