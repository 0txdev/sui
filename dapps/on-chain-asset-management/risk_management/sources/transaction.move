module risk_management::transaction {

    use risk_management::policy_config::{Self, SpenderCap, ApproverCap, Assets};
    use sui::tx_context::{Self, TxContext};
    use sui::object::{Self, UID, ID};
    use sui::coin::{Self};
    use std::string::{Self, String};
    use sui::transfer;
    use std::vector as vec;

    const EAmountLimitExceeded : u64 = 0;
    const ETimeLimitExceeded : u64 = 1;
    const EAlreadyApprovedByThisApprover : u64 = 2;
    const ENotOriginalOwnerOfCapability : u64 = 3;
    const ESpenderCannotSendMoneyToHimself : u64 = 4;

    struct TransactionRequest has key, store {
        id: UID,
        amount: u64,
        time_limit: u64,
        spender: address,
        recipient: address,
        description: String,
        approvers_num: u64,
        approved_by: vector<address>,
    }

    struct TransactionApproval has key, store {
        id: UID,
        transaction_id: ID,
        amount: u64,
        spender: address,
        recipient: address,
        approvers: vector<address>,
    }

    entry fun initiate_transaction(
        spender_cap: &SpenderCap,
        amount: u64,
        recipient: address,
        description: vector<u8>,
        ctx: &mut TxContext,
    ) { 
        assert!(policy_config::get_spender_original_owner(spender_cap) == tx_context::sender(ctx), 3);
        assert!(recipient != tx_context::sender(ctx), 4);
        assert!(amount <= policy_config::get_amount_limit(spender_cap), 0);
        transfer::share_object(TransactionRequest {
            id: object::new(ctx),
            amount,
            time_limit: policy_config::get_time_limit(spender_cap) + tx_context::epoch(ctx),
            spender: tx_context::sender(ctx),
            recipient,
            description: string::utf8(description),
            approvers_num: policy_config::get_approvers_num(spender_cap),
            approved_by: vec::empty()
        })
    }

    entry fun approve_request(
        approver_cap: &ApproverCap,
        tx_request: &mut TransactionRequest,
        ctx: &mut TxContext,
    ) {
        assert!(policy_config::get_approver_original_owner(approver_cap) == tx_context::sender(ctx), 3);
        assert!(tx_context::epoch(ctx) <= tx_request.time_limit, 1);
        assert!(vec::contains(&tx_request.approved_by, &tx_context::sender(ctx)) == false, 2);
        if (vec::length(&tx_request.approved_by) < tx_request.approvers_num - 1) {
            vec::push_back(&mut tx_request.approved_by, tx_context::sender(ctx))
        } else {
            vec::push_back(&mut tx_request.approved_by, tx_context::sender(ctx));
            let newvector : vector<address> = tx_request.approved_by;
            transfer::transfer(TransactionApproval{
                id: object::new(ctx),
                transaction_id: object::uid_to_inner(&tx_request.id),
                amount: tx_request.amount,
                spender: tx_request.spender,
                recipient: tx_request.recipient,
                approvers: newvector,
            }, tx_request.spender);
        }
    }

    entry fun reject_request(
        _: &ApproverCap,
        _tx_request: &TransactionRequest,
        _ctx: &mut TxContext,
    ) {
        // An event should be emitted informing that the request with
        // transaction_id: object::uid_to_inner(&tx_request.id) got rejected
    }

    entry fun execute_transaction(
        spender_cap: &mut SpenderCap,
        tx_approval: TransactionApproval,
        assets: &mut Assets,
        ctx: &mut TxContext,
    ) {
        assert!(policy_config::get_spender_original_owner(spender_cap) == tx_context::sender(ctx), 3);
        assert!(tx_approval.amount <= policy_config::get_amount_limit(spender_cap), 0);
        transfer::transfer(
            coin::take(policy_config::get_foundation_balance_mut(assets), tx_approval.amount, ctx) , 
            tx_approval.recipient
        );
        
        //Unpack and delete the TransactionApproval
        let TransactionApproval { 
            id,
            transaction_id: _,
            amount: _,
            spender: _,
            recipient: _,
            approvers: _ } = tx_approval;
        object::delete(id)
    }

}