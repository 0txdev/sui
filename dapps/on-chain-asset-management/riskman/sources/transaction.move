module riskman::transaction {

    use riskman::policy_config::{Self, SpenderCap, ApproverCap, Assets};
    use sui::tx_context::{Self, TxContext};
    use sui::object::{Self, UID, ID};
    use sui::coin::{Self};
    use std::string::{Self, String};
    use sui::transfer;

    const EAmountLimitExceeded : u64 = 0;
    const ETimeLimitExceeded : u64 = 1;
    const EAmountLimitExceededAfterApproval : u64 = 2;

    struct TransactionRequest has key, store {
        id: UID,
        amount: u64,
        time_limit: u64,
        spender: address,
        recipient: address,
        description: String,
    }

    struct TransactionApproval has key, store {
        id: UID,
        transaction_id: ID,
        amount: u64,
        spender: address,
        recipient: address,
        approver: address,
    }

    entry fun initiate_transaction(
        spender_cap: &SpenderCap,
        amount: u64,
        recipient: address,
        description: vector<u8>,
        ctx: &mut TxContext,
    ) { 
        assert!(amount <= policy_config::get_amount_limit(spender_cap) - policy_config::get_amount_spent(spender_cap), 0);
        transfer::share_object(TransactionRequest {
            id: object::new(ctx),
            amount,
            time_limit: policy_config::get_time_limit(spender_cap) + tx_context::epoch(ctx),
            spender: tx_context::sender(ctx),
            recipient,
            description: string::utf8(description)
        })
    }

    entry fun approve_request(
        _: &ApproverCap,
        tx_request: &TransactionRequest,
        ctx: &mut TxContext,
    ) {
        assert!(tx_context::epoch(ctx) <= tx_request.time_limit, 1);
        transfer::transfer(TransactionApproval{
            id: object::new(ctx),
            transaction_id: object::uid_to_inner(&tx_request.id),
            amount: tx_request.amount,
            spender: tx_request.spender,
            recipient: tx_request.recipient,
            approver: tx_context::sender(ctx)
        }, tx_request.spender);
    }

    entry fun reject_request(
        _: &ApproverCap,
        _tx_request: &TransactionRequest,
        _ctx: &mut TxContext,
    ) {
        // An event should be emitted informing that the request with
        // transaction_id: object::uid_to_inner(&tx_request.id) got rejected
    }

    entry fun execute_transaction(
        spender_cap: &mut SpenderCap,
        tx_approval: TransactionApproval,
        assets: &mut Assets,
        ctx: &mut TxContext,
    ) {
        assert!(tx_approval.amount <= policy_config::get_amount_limit(spender_cap) - policy_config::get_amount_spent(spender_cap), 2);
        transfer::transfer(
            coin::take(policy_config::get_foundation_balance_mut(assets), tx_approval.amount, ctx) , 
            tx_approval.recipient
        );
        policy_config::update_spent(spender_cap, tx_approval.amount);
        
        //Unpack and delete the TransactionApproval
        let TransactionApproval { 
            id,
            transaction_id: _,
            amount: _,
            spender: _,
            recipient: _,
            approver: _ } = tx_approval;
        object::delete(id)
    }

}