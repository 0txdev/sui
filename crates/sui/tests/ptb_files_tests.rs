// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeMap, path::Path};
use sui::client_ptb::{ptb::PTB, ptb_builder::errors::render_errors};
use sui_types::transaction::{CallArg, ObjectArg};
use test_cluster::TestClusterBuilder;

const TEST_DIR: &str = "tests";

#[cfg(not(msim))]
#[tokio::main]
async fn test_ptb_files(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let _ = miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .color(false)
                .width(80)
                .build(),
        )
    }));

    let fname = || path.file_name().unwrap().to_string_lossy().to_string();
    let file = path.to_str().unwrap().to_string();
    let args = vec!["--file".to_string(), file, "--preview".to_string()];
    let cwd = std::env::current_dir().unwrap();
    let mut commands = Vec::new();
    let ptb = PTB { args: args.clone() };
    ptb.parse_args(cwd, args, &mut BTreeMap::new(), &mut commands)?;

    // === PREVIEW ===
    let ptb_preview = ptb.preview(&commands.clone());
    let mut results = vec![];
    let preview_string = if let Some(ptb_preview) = ptb_preview {
        ptb_preview.to_string()
    } else {
        "".to_string()
    };
    results.push(" === PREVIEW === ".to_string());
    results.push(preview_string);

    // === PARSE COMMANDS ===
    let parsed_ptb_commands = ptb.parse_ptb_commands(commands.clone());
    match parsed_ptb_commands {
        Ok(ref commands) => commands,
        Err(e) => {
            // insta::assert_display_snapshot!(
            //     fname(),
            //     format!("=== FILE EXPANSION ERROR === \n {:?}", e)
            // );
            return Ok(());
        }
    };
    let parsed_ptb_commands = match parsed_ptb_commands {
        Ok(parsed) => parsed,
        Err(errors) => {
            let rendered = render_errors(&commands, errors);
            results.push(" === ERRORS AFTER PARSING INPUT COMMANDS === ".to_string());
            for e in rendered.iter() {
                results.push(format!("{:?}", e));
            }
            insta::assert_display_snapshot!(fname(), results.join("\n"));
            return Ok(());
        }
    };

    results.push(" === PARSED INPUT COMMANDS === ".to_string());

    for (_, c) in &parsed_ptb_commands {
        results.push(format!("cmd: {:?}", c));
    }

    // === BUILD PTB ===
    let test_cluster = TestClusterBuilder::new().build().await;
    let context = test_cluster.wallet;
    let client = context.get_client().await?;

    let built_ptb = ptb
        .parse_and_build_ptb(parsed_ptb_commands, &context, client)
        .await;

    if let Ok(ref ptb) = built_ptb {
        results.push(" === BUILT PTB === ".to_string());
        for (i, ca) in ptb.0.inputs.iter().enumerate() {
            results.push(format!("Input {}: {}", i, stable_call_arg_display(ca)));
        }
        for (i, c) in ptb.0.commands.iter().enumerate() {
            results.push(format!("Command {}: {}", i, c));
        }
    }

    // === BUILDING PTB ERRORS ===
    if let Err(e) = built_ptb {
        let rendered = render_errors(&commands, e);

        results.push(" === BUILDING PTB ERRORS === ".to_string());
        for e in rendered.iter() {
            results.push(format!("{:?}", e));
        }
    }

    // === FINALLY DO THE ASSERTION ===
    insta::assert_display_snapshot!(fname(), results.join("\n"));

    Ok(())
}

fn stable_call_arg_display(ca: &CallArg) -> String {
    match ca {
        CallArg::Pure(v) => format!("Pure({:?})", v),
        CallArg::Object(oa) => match oa {
            ObjectArg::ImmOrOwnedObject(_) => "ImmutableOrOwnedObject".to_string(),
            ObjectArg::SharedObject { mutable, .. } => {
                format!("SharedObject(mutable: {})", mutable)
            }
            ObjectArg::Receiving(_) => "Receiving".to_string(),
        },
    }
}

datatest_stable::harness!(test_ptb_files, TEST_DIR, r".*\.ptb$",);
