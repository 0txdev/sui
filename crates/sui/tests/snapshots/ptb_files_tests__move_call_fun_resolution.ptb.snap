---
source: crates/sui/tests/ptb_files_tests.rs
expression: "results.join(\"\\n\")"
---
 === PREVIEW === 
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ PTB Preview                                                                                                  │
├────────────┬────────────────────────────────────────────────────────┬────────────────────────────────────────┤
│ command    │ from                                                   │ value(s)                               │
├────────────┼────────────────────────────────────────────────────────┼────────────────────────────────────────┤
│ gas_budget │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ 1000                                   │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ zero_x_zero::option::is_none<u64> none │
│ assign     │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ a none                                 │
│ assign     │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ zero_x_zero @0x0                       │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ zero_x_zero::option::is_none<u64> a    │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ std::Option::is_none<u64> a            │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ std::option::isnone<u64> a             │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ std::option::is_none<u64>              │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ std::option::is_none<u64> a b          │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ sui::coin::destroy_zero @0x1           │
│ move_call  │ tests/ptb_files/move_call/move_call_fun_resolution.ptb │ std::option::is_none<u64, bool> a      │
│ preview    │ console                                                │ true                                   │
╰────────────┴────────────────────────────────────────────────────────┴────────────────────────────────────────╯
 === PARSED INPUT COMMANDS === 
cmd: GasBudget(Spanned { span: Span { start: 0, end: 4, arg_idx: 0, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: 1000 })
cmd: MoveCall(Spanned { span: Span { start: 0, end: 28, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("zero_x_zero") }, module_name: Spanned { span: Span { start: 13, end: 19, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 21, end: 28, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 28, end: 33, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [Spanned { span: Span { start: 0, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 0, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: None }) }])
cmd: Assign(Spanned { span: Span { start: 0, end: 1, arg_idx: 0, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: "a" }, Some(Spanned { span: Span { start: 0, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Option(Spanned { span: Span { start: 0, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: None }) }))
cmd: Assign(Spanned { span: Span { start: 0, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: "zero_x_zero" }, Some(Spanned { span: Span { start: 1, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Address(0x0) }))
cmd: MoveCall(Spanned { span: Span { start: 0, end: 28, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("zero_x_zero") }, module_name: Spanned { span: Span { start: 13, end: 19, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 21, end: 28, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 28, end: 33, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [Spanned { span: Span { start: 0, end: 1, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("a") }])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("std") }, module_name: Spanned { span: Span { start: 5, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("Option") }, function_name: Spanned { span: Span { start: 13, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 20, end: 25, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [Spanned { span: Span { start: 0, end: 1, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("a") }])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 19, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("std") }, module_name: Spanned { span: Span { start: 5, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 13, end: 19, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("isnone") } } }, Some(Spanned { span: Span { start: 19, end: 24, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [Spanned { span: Span { start: 0, end: 1, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("a") }])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("std") }, module_name: Spanned { span: Span { start: 5, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 13, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 20, end: 25, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("std") }, module_name: Spanned { span: Span { start: 5, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 13, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 20, end: 25, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64] }), [Spanned { span: Span { start: 0, end: 1, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("a") }, Spanned { span: Span { start: 2, end: 3, arg_idx: 1, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("b") }])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 23, arg_idx: 0, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("sui") }, module_name: Spanned { span: Span { start: 5, end: 9, arg_idx: 0, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("coin") }, function_name: Spanned { span: Span { start: 11, end: 23, arg_idx: 0, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("destroy_zero") } } }, None, [Spanned { span: Span { start: 1, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 9, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Address(0x1) }])
cmd: MoveCall(Spanned { span: Span { start: 0, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: ModuleAccess { address: Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Named("std") }, module_name: Spanned { span: Span { start: 5, end: 11, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("option") }, function_name: Spanned { span: Span { start: 13, end: 20, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("is_none") } } }, Some(Spanned { span: Span { start: 20, end: 31, arg_idx: 0, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: [U64, Bool] }), [Spanned { span: Span { start: 0, end: 1, arg_idx: 1, file_scope: FileScope { file_command_index: 10, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Identifier("a") }])
cmd: Preview(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 11, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Bool(false) })
cmd: WarnShadows(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 12, name: "tests/ptb_files/move_call/move_call_fun_resolution.ptb", name_index: 0 } }, value: Bool(false) })
cmd: Preview(Spanned { span: Span { start: 0, end: 4, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "console", name_index: 0 } }, value: Bool(true) })
cmd: WarnShadows(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 2, name: "console", name_index: 0 } }, value: Bool(false) })
 === BUILDING PTB ERRORS === 
  × Error at command 1 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call zero_x_zero::option::is_none<u64> none 
   ·           ─────┬─────
   ·                ╰── Unbound named address: 'zero_x_zero'
   ╰────
  help: This is most likely because the named address 'zero_x_zero' is not
        in scope. You can either bind a variable to the address that you
        want to use or use the address in the command.

  × Error at command 4 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call zero_x_zero::option::is_none<u64> a 
   ·           ─────┬─────
   ·                ╰── Object 0x0000000000000000000000000000000000000000000000000000000000000000 does not exist.
   ╰────

  × Error at command 5 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call std::Option::is_none<u64> a 
   ·                ───┬──
   ·                   ╰── Module not found in package: "Option".
   ╰────
  help: Did you mean 'option'?

  × Error at command 6 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call std::option::isnone<u64> a 
   ·                        ───┬──
   ·                           ╰── Could not resolve function 'isnone' in module 'option'
   ╰────
  help: Did you mean 'is_none'?

  × Error at command 7 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call std::option::is_none<u64> 
   ·           ──────────┬─────────
   ·                     ╰── Expected 1 arguments, but got 0
   ╰────

  × Error at command 8 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call std::option::is_none<u64> a b 
   ·                                     ─┬─
   ·                                      ╰── Expected 1 arguments, but got 2
   ╰────

  × Error at command 9 in PTB file 'tests/ptb_files/move_call/
  │ move_call_fun_resolution.ptb'
   ╭────
 1 │ move_call sui::coin::destroy_zero @0x1 
   ·                                    ─┬─
   ·                                     ╰── Not enough type parameters supplied for Move call
   ╰────

